src/plugins/data/common/search/expressions/esql.ts

```diff
@@ -238,37 +237,28 @@ export const getEsqlFn = ({ getStartDependencies }: EsqlFnArguments) => {
           );
         }),
         map(({ rawResponse: body, warning }) => {
-          const columns =
-            body.columns?.map(({ name, type }) => ({
-              id: name,
-              name,
-              meta: { type: normalizeType(type) },
-            })) ?? [];
           // all_columns in the response means that there is a separation between
           // columns with data and empty columns
           // columns contain only columns with data while all_columns everything
           const hasEmptyColumns =
             body.all_columns && body.all_columns?.length > body.columns.length;
+          const lookup = new Set(
+            hasEmptyColumns ? body.columns?.map(({ name }) => name) || [] : []
+          );
+          const allColumns =
+            (body.all_columns ?? body.columns)?.map(({ name, type }) => ({
+              id: name,
+              name,
+              meta: { type: normalizeType(type) },
+              isNull: hasEmptyColumns ? !lookup.has(name) : false,
+            })) ?? [];
 
-          let emptyColumns: DatatableColumn[] = [];
-
+          // sort only in case of empty columns
           if (hasEmptyColumns) {
-            const difference =
-              body.all_columns?.filter((col1) => {
-                return !body.columns.some((col2) => {
-                  return col1.name === col2.name;
-                });
-              }) ?? [];
-            emptyColumns =
-              difference?.map(({ name, type }) => ({
-                id: name,
-                name,
-                meta: { type: normalizeType(type) },
-                isNull: true,
-              })) ?? [];
+            allColumns.sort((a, b) => Number(a.isNull) - Number(b.isNull));
```

Comment:
Why do we sort? I think in UI it's ordered by name anyway.

Reply:
Because with ?drop_null_columns you get only the values from non empty columns.

Reply:
Yes but what would not work without sorting on `allColumns`?

Reply:
No because you might have:

```
[{colA: null}, {colB: not null}, {colC: null},  {colD: not null}]
```

but on the values you get only

```
[1, 'a'] --> 1 is the value of col B and 'a' the value of colD
```

so if you don't put the null columns in the end then you are not having a correct match between your columns and values.

Reply:
Got it. Do we have a test to make sure that values are read correctly according to columns?

Reply:
Yes I added one in my PR (the one we merged)